tests/core/full_node/test_tx_processing_queue.py

```diff
@@ -0,0 +1,184 @@
+        await transaction_queue.put(tx, None)
+
+    resulting_txs = []
+    for _ in range(num_txs + len(local_txs)):
+        resulting_txs.append(await transaction_queue.pop())
+
+    for i in range(num_txs + len(local_txs)):
+        if i < len(local_txs):
+            assert local_txs[i] == resulting_txs[i]  # first 20 should come from local
+        else:
+            assert list_txs[i - 20] == resulting_txs[i]
+
+    # now we validate that the pop command is blocking
+    task = asyncio.create_task(transaction_queue.pop())
+    with pytest.raises(asyncio.InvalidStateError):  # task is not done, so we expect an error when getting result
+        task.result()
+    # add a tx to test task completion
+    await transaction_queue.put(get_transaction_queue_entry(None, 0), None)
+    await asyncio.wait_for(task, 1)  # we should never time out here
+
+
+@pytest.mark.asyncio
+async def test_lots_of_peers() -> None:
+    transaction_queue = TransactionQueue(1000, log)
+    num_peers = 1000
+    num_txs = 100
+    total_txs = num_txs * num_peers
+    peer_ids: List[bytes32] = [get_peer_id() for _ in range(num_peers)]
+
+    # 100 txs per peer
+    list_txs = [get_transaction_queue_entry(peer_id, i) for peer_id in peer_ids for i in range(num_txs)]
+    for tx in list_txs:
+        await transaction_queue.put(tx, tx.peer_id)  # type: ignore[attr-defined]
+
+    assert transaction_queue._index_to_peer_map == peer_ids  # make sure all peers are in the map
+
+    resulting_txs = []
+    for _ in range(total_txs):
+        resulting_txs.append(await transaction_queue.pop())
+
+    assert transaction_queue._index_to_peer_map == []  # we should have removed all the peer ids
+    # There are 1000 peers, so each peer will have one transaction processed every 1000 iterations.
+    for i in range(num_txs):
+        assert list_txs[i] == resulting_txs[i * 1000]
+
+
+@pytest.mark.asyncio
+async def test_full_queue() -> None:
+    transaction_queue = TransactionQueue(1000, log)
+    num_peers = 100
+    num_txs = 1000
+    total_txs = num_txs * num_peers
+    peer_ids: List[bytes32] = [get_peer_id() for _ in range(num_peers)]
+
+    # 999 txs per peer then 1 to fail later
+    list_txs = [get_transaction_queue_entry(peer_id, i) for peer_id in peer_ids for i in range(num_txs)]
+    for tx in list_txs:
+        await transaction_queue.put(tx, tx.peer_id)  # type: ignore[attr-defined]
+
+    assert transaction_queue._index_to_peer_map == peer_ids  # make sure all peers are in the map
+
+    # test failure case.
+    with pytest.raises(TransactionQueueFull):
+        await transaction_queue.put(get_transaction_queue_entry(peer_ids[0], 1001), peer_ids[0])
+
+    resulting_txs = []
+    for _ in range(total_txs):
+        resulting_txs.append(await transaction_queue.pop())
+
+    assert transaction_queue._index_to_peer_map == []  # we should have removed all the peer ids
+
+
+@pytest.mark.asyncio
+async def test_queue_cleanup_and_fairness() -> None:
```

Comment:
this test could be as simple as:

* pushing 2 entries for peer A, one entry for B and 2 entries for C
* pop 3 items (ensuring it's from A, B and C)
* pop 2 items, ensuring it's from A and C (in that order)

It's problematic to have tests assert state of implementation details. One of the purposes of tests is to ensure the class behaves correctly after refactoring it, or changing the implementation strategy. With tests like this, one would have to re-write the tests as well, which defeats an important point of having them.